clang/docs/checkedc/verification-with-seahorn.md

```diff
@@ -0,0 +1,96 @@
+
+For example, for this bounds information: `int *a : count(n)`, we will have: `assume(a != NULL); assume(n >= 0);`. The assumption on function calls having bounds-safe interface is needed to be able to have enough bounds information to propagate over the pointer assignments.
+
+## Installing Seahorn
+
+The following steps are quick listing for installing seahorn. For more detailed instructions please refer to Seahorn [github page](https://github.com/seahorn/seahorn).
+
+1. `git clone https://github.com/seahorn/seahorn.git`
+2. `cd seahorn; mkdir build; cd build`
+3. `cmake -DCMAKE_INSTALL_PREFIX=run -DCMAKE_BUILD_TYPE=Release ..`
+4. `cmake --build .`
+5. `cmake --build . --target extra && cmake ..`
+6. `cmake --build . --target crab && cmake ..`
+7. `cmake --build . --target install`
+
+The executables and scripts will be in `seahorn/build/run/bin`.
+
+## Running Seahorn
+
+The main script to launch seahorn and configuring its pipelines is `sea`. You can see the available command by `sea -h`. The main verification pipeline can be run with `sea pf file.c`. This is a shortcut for running the _Front End_ and _Solving Horn constraints_ (`fe | horn --solve`). The output will be either `sat` or `unsat`. `unsat` means that starting from the safe assumptions, there is no way for the program to get to bad states, defined by assertions (assertions will not fail). In other words, the program is safe. For example, for the code below, that is annotated with the assumptions and assertions, running `sea pf file.c` will return `unsat`.
+```c
+#include "seahorn/seahorn.h"
+extern int non_deterministic_value();
+
+int main() {
+    int x = 1, y = 1;
+    int x_old = x, y_old = y;
+    int n = non_deterministic_value();
+    while (n--) {
+        int t1 = x;
+        int t2 = y;
+        x = t1 + 1;
+        y = t2 - 1;
+    }
+    sassert(x_old + y_old == x + y);
+    return 0;
+}
+```
+
+
+## Checked C and Seahorn
+
+As mentioned above, we replace the front-end of Seahorn with Checked C, to use the bounds information and inject necessary assertion for the verification. The `sea fe` (Front End) is a shortcut for _clang_, _Pre processing_, _Mixed Semantics_ and _Compiler Optimization_ (`clang | pp | ms | opt`).
+
+Checked C will inject calls to verifier functions when `-finject-verifier-calls` is given. Note that this flag will generate a bitcode that does not link, because the calls do not have definitions. To enable adding bounds and consequently dynamic checks for unchecked pointers, `-funchecked-pointers-dynamic-check` flag should be given.
+
+We simply use Checked C clang instead of clang in the front-end pipeline to have the desired verification conditions. The following script shows an example pipeline. Each of the steps can be further configured with command line arguments.
+```bash
+INC= # /path/to/seahorn/build/run/include
+clang -c -emit-llvm -finject-verifier-calls -funchecked-pointers-dynamic-check -O0 -m64 -I $INC file.c -o file.bc

```

Comment:
Is there a simpler way to invoke Seahorn with the Checked C clang compiler as the front-end?  Or do people always need to use these 4 steps?

Reply:
I added a comment on how to only run `sea pf file.c`.